### DISCUSSION TITLE
Welcome to Tetra Discussions!

### DISCUSSION THREAD

**Discussion by 17cupsofcoffee** [17cupsofcoffee/tetra#217] [General]
## 👋 Welcome!
I've finally been given access to the Discussions beta, so I thought it'd be fun to try it out.

The [issue tracker](https://github.com/17cupsofcoffee/tetra/issues) is still the best place to post your bug reports or concrete feature requests, but this message board is a good place for:

* Asking questions about how to use Tetra (or how to use it more effectively)
* Discussing future developments to the framework, or things that you think are missing
* Posting cool stuff you've made with Tetra

We'll see how it goes for a few weeks - if it gets a decent amount of use, I'll keep it, if not, stuff can easily be migrated to the issue tracker :)

**Comment by sumibi-yakitori** [comment]
Thanks for the nice framework.
How do you yourself want to develop tetra?

**Reply by 17cupsofcoffee** [reply]
> Thanks for the nice framework.

Thank you for using it, and the PRs :)

> How do you yourself want to develop tetra?

Good question, I need to write up a roadmap somewhere...

The main things I'm thinking of doing this year are:

* Rewrite the audio layer using [`oddio`](https://github.com/Ralith/oddio), once that gets released.
    * It seems much more well thought out than `rodio` and should allow me to improve the audio API a bit.
* Add support for [bitmap fonts](https://www.angelcode.com/products/bmfont/) and/or switch to a TTF rendering library that actually lets you disable anti-aliasing.
    * Drawing text in low-resolution/pixel art games doesn't work very well at the moment, and that's a real problem for the sort of stuff I want to use Tetra for...
* Rethink the `Drawable` API a bit.
    * See #94 - it's hard to write a `Drawable` trait that works for every possible object. Also, the fact that it takes `&self` instead of `&mut self` really complicates the `Text` implementation.
* *Maybe* add support for primitive shape drawing.
    * I'm kinda tempted to just ask people to use [`lyon`](https://github.com/nical/lyon) and `tetra::graphics::Mesh` to implement that themselves though, as it's not something I'd personally use at all.
* *Maybe* try find a way to restructure the API to be a little less stateful, so that it's easier to build stuff on top of Tetra.
    * e.g. if you have a library that renders stuff using a custom shader and matrix, it kinda sucks that you have to reset all that state before you pass control back. I don't have a concrete idea of how to solve this yet, though...
* Probably some other stuff that I'm forgetting or haven't thought of yet. 

In general, though, I'm going to try to avoid adding any *really* big new features until I (or other people) actually have a project that requires them. I'd like to actually *use* Tetra this year, rather than just endlessly tinkering with it 😅

**Reply by sumibi-yakitori** [reply]
Hmmm...I see.
Like you, I don't think I'll be using primitives myself at all😜

As for the font renderer that can disable anti-aliasing,
I've been wanting to make one for a long time, if I had the time.

After trying Bevy and ggez
The reason why I finally decided to use tetra was because I could feel the love for 2D games in your examples.
I also wanted to learn the structure of MonoGame, XNA, FNA that made Celeste.
Unfortunately, in the country where I live, I can hardly find anyone releasing games using those solid and thin game frameworks.

Because you are making music
I feel like someone who is passionate about expressing something.
I'm looking forward to the game you make.
Are you going to make a game like Fire Emblem?

I wish I could use Rust to make games, but the game I'm working on now has a goal to be released on Nintendo Switch, so I have to compromise and use Unity.

**Reply by 17cupsofcoffee** [reply]
> The reason why I finally decided to use tetra was because I could feel the love for 2D games in your examples.

This means a lot to me! The most fun I've had doing gamedev was playing around with 2D engines like XNA and [Otter](http://otter2d.com/), so Tetra is basically just me trying to capture that same feeling in Rust 😄 

> I also wanted to learn the structure of MonoGame, XNA, FNA that made Celeste.

If you're interested in Celeste (also one of my favourite games in recent years!), you may also want to look at:
* The [Celeste GitHub](https://github.com/NoelFB/Celeste) - the `Player` class and the original PICO-8 prototype have been open-sourced, and they're interesting to look at.
* [This blog post on the physics/collision detection](https://medium.com/@MattThorson/celeste-and-towerfall-physics-d24bd2ae0fc5).
* [Strawberry](https://github.com/MaddyThorson/Strawberry), a game engine Maddy Thorson (one of the Celeste devs) has been writing using a language called [Beef](https://www.beeflang.org/). The example game implements the physics from the previously mentioned blog post.
* [blah](https://github.com/NoelFB/blah/), a game engine Noel Berry (the other Celeste dev) has been writing in C++.

I've been [playing around with implementing something based on those physics](https://twitter.com/17cupsofcoffee/status/1348287675942502401) recently as a learning exercise 🙂

> Are you going to make a game like Fire Emblem?

I'm not sure yet! I've been prototyping a Fire Emblem-ish game with Tetra, but it seems like a massive undertaking to build something like that - thinking of doing some game jams/smaller games before diving into any really big projects.

**Reply by sumibi-yakitori** [reply]
> This means a lot to me! The most fun I've had doing gamedev was playing around with 2D engines like XNA and Otter, so Tetra is basically just me trying to capture that same feeling in Rust 😄

Come to think of it, it would look more appealing if you put a screenshot of your example in README.md 😄

> If you're interested in Celeste (also one of my favourite games in recent years!), you may also want to look at:

I love Celeste too.

Wow.
I knew to some extent that Maddy Thorson and Noel Berry were willing to share their expertise with the public, but I didn't know they both made game engines.
What a powerful development team they are.
How do English-speaking people build such a strong team to make indie games? 😂 

> I'm not sure yet! I've been prototyping a Fire Emblem-ish game with Tetra, but it seems like a massive undertaking to build something like that - thinking of doing some game jams/smaller games before diving into any really big projects.

That's good 👍

**Reply by BoxyUwU** [reply]
Haha I've been thinking of making a celeste inspired platformer with tetra for a while now what a coincedence xD
